[ao][sparsity] Added the Nearly Diagonal Sparsifier

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78448

This sparsifier creates a nearly diagonal mask to be applied to the weight matrix.
    Nearly Diagonal Matrix is a matrix that contains non-zero elements near the diagonal and the rest are zero.
    An example of a nearly diagonal matrix with degree (or nearliness) 3 and 5 are follows respectively.
    1 1 0 0       1 1 1 0
    1 1 1 0       1 1 1 1
    0 1 1 1       1 1 1 1
    0 0 1 1       0 1 1 1
    Note that a nearly diagonal matrix with degree 1 is just a matrix with main diagonal populated

    This sparsifier is controlled by one variable:
    1. `nearliness` defines the number of non-zero diagonal lines that are closest to the main diagonal.
        Currently - supports only odd number

    Note:
        This can be accelerated (vectorized) once the Spdiagonal feature (PR: #78439) is landed or the banded matrix
        feature is landed: https://stackoverflow.com/questions/52463972/generating-banded-matrices-using-numpy

Test Plan:

```
python test/test_ao_sparsity.py TestNearlyDiagonalSparsifier
```

Differential Revision: [D36905916](https://our.internmc.facebook.com/intern/diff/D36905916)

Commit:
Added the Nearly Diagonal Sparsifier

[ghstack-poisoned]

vadimkantorov commented on torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:
```diff
@@ -0,0 +1,51 @@
+            raise ValueError("nearliness cannot be larger than the "
+                            "dimensions of weight matrix.")
+        # create mask
+        for row in range(0, height):
```
just curious, can this be done in a vectorized way? (maybe a good moment to introduce a toeplitz matrix constructor? some options from numpy at https://stackoverflow.com/questions/52463972/generating-banded-matrices-using-numpy)

macandro96 replied:
Thanks for the suggestion. We could do this by identifying the entries that needs to be set and just fill those entries.  Have implemented this. But not sure how we could use a Toeplitz matrix to accomplish this.

macandro96 replied:
Correct

vadimkantorov replied:
So a toeplitz matrix or banded matrix constructor should be helpful for this

z-a-f replied:
@vadimkantorov Because there is no toeplitz constructor kernel implemented, do you think it's fine to put it as a comment in the docstring, and leave it as is?

vadimkantorov replied:
yep, I think so! then if it's implemented, this code can be updated

macandro96 replied:
Added the comment in doc-string. Will update once implemented.

Commit:
Update on "Added the Nearly Diagonal Sparsifier"




[ghstack-poisoned]

vadimkantorov commented on torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:
```diff
@@ -0,0 +1,52 @@
+            # Bounds of entries that needs to be set to 1
+            low = max(0, row - dist_to_diagonal)
+            high = min(width, row + dist_to_diagonal + 1)
+            mask[row, low:high] = torch.ones(high - low, device=mask.device)
```
could it just be `mask[row, low:high].fill_(1)`?

macandro96 replied:
That makes sense. I'll amend my commit and raise a pr again.

Commit:
Update on "Added the Nearly Diagonal Sparsifier"




[ghstack-poisoned]

HDCharles commented on test/ao/sparsity/test_sparsifier.py:
```diff
@@ -346,3 +346,119 @@ def test_sparsity_levels(self):
+
+
+# helper function to verify nearliness of a mask
+def _verify_nearliness(mask: torch.Tensor, nearliness: int):
```
we tend to put these helpers in the test class

macandro96 replied:
Got it!

HDCharles reviewed (commented):
should include tags [ao][sparsity] in the PR title

Commit:
Update on "Added the Nearly Diagonal Sparsifier"




[ghstack-poisoned]

albanD commented on test/allowlist_for_publicAPI.json:
```diff
@@ -514,6 +514,9 @@
     "Tuple",
     "reduce"
   ],
+  "torch.ao.sparsity.sparsifier.nearly_diagonal_sparsifier": [
```
You should not add new entries to this file.
Make sure to follow the guidelines from the error message to make this properly public or private depending on what it should be.

z-a-f commented on torch/ao/sparsity/__init__.py:
```diff
@@ -5,6 +5,7 @@
 # Sparsifier
 from .sparsifier.base_sparsifier import BaseSparsifier
 from .sparsifier.weight_norm_sparsifier import WeightNormSparsifier
+from .sparsifier.nearly_diagonal_sparsifier import NearlyDiagonalSparsifier
```
Add the following in this file:

```python
__all__ = [
    'NearlyDiagonalSparsifier',
]
```

I think that should fix the issue. I will all other imports in there in the later PRs. cc @albanD

z-a-f replied:
Actually scratch that -- We probs need to avoid exposing the `BaseSparsifier` in the `nearly_diagonal...` file

Commit:
Update on "Added the Nearly Diagonal Sparsifier"




[ghstack-poisoned]

z-a-f reviewed (approved):
LGTM

Commit:
Update on "[ao][sparsity]Added the Nearly Diagonal Sparsifier"




[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/sparsifier/nearly_diagonal_sparsifier.py:
```diff
@@ -0,0 +1,55 @@
+        if nearliness <= 0:
+            return
+
+        weight = layer.weight
+        height, width = weight.shape
```
nit:
```
height, width = layer.weight.shape
```

z-a-f commented:
FYI, I checked different ways of implementing the logic, and it looks like vectorization might give a slight edge: https://colab.research.google.com/drive/1USyM4mTuAFml5ntNX-TuUPVtY7AJDtAB?usp=sharing

I won't block it on this, just an fyi :)

HDCharles reviewed (approved):
looks good though there's a failing CI test that you need to resolve before landing, not sure what here could be causing it though, probably just rebase and see if it goes away.

Commit:
Update on "[ao][sparsity]Added the Nearly Diagonal Sparsifier"




[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36905916).

Commit:
Update on "[ao][sparsity] Added the Nearly Diagonal Sparsifier"


This sparsifier creates a nearly diagonal mask to be applied to the weight matrix.
    Nearly Diagonal Matrix is a matrix that contains non-zero elements near the diagonal and the rest are zero.
    An example of a nearly diagonal matrix with degree (or nearliness) 3 and 5 are follows respectively.
    1 1 0 0       1 1 1 0
    1 1 1 0       1 1 1 1
    0 1 1 1       1 1 1 1
    0 0 1 1       0 1 1 1
    Note that a nearly diagonal matrix with degree 1 is just a matrix with main diagonal populated

    This sparsifier is controlled by one variable:
    1. `nearliness` defines the number of non-zero diagonal lines that are closest to the main diagonal.
        Currently - supports only odd number

    Note:
        This can be accelerated (vectorized) once the Spdiagonal feature (PR: #78439) is landed or the banded matrix
        feature is landed: https://stackoverflow.com/questions/52463972/generating-banded-matrices-using-numpy

Test Plan:

```
python test/test_ao_sparsity.py TestNearlyDiagonalSparsifier
```

Differential Revision: [D36905916](https://our.internmc.facebook.com/intern/diff/D36905916)

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36905916).

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36905916).
